File: src/aib.c
/*
Copyright (C) 2007-12 Andrea Vedaldi and Brian Fulkerson.
All rights reserved.

This file is part of the VLFeat library and is made available under
the terms of the BSD license (see the COPYING file).
*/

#include <stdio.h>
#include <stdlib.h>
#include <vl/aib.h>

int main()
{
    vl_uint32 nrows = 10;
    vl_uint32 ncols = 3;
    double Pic[3*10] = {
    0.6813,    0.3028,    0.8216,
    0.3795,    0.5417,    0.6449,
    0.8318,    0.1509,    0.8180,
    0.5028,    0.6979,    0.6602,
    0.7095,    0.3784,    0.3420,
    0.4289,    0.8600,    0.2897,
    0.3046,    0.8537,    0.3412,
    0.1897,    0.5936,    0.5341,
    0.1934,    0.4966,    0.7271,
    0.6822,    0.8998,    0.3093,
    };

    vl_uint32 r,c;
    VlAIB * aib;
    vl_uint * parents;

    printf("Pic = [");
    for(r=0; r<nrows; r++)
    {
        for(c=0; c<ncols; c++)
            printf("%f ", Pic[r*ncols+c]);
        printf("; ...\n");
    }
    printf("];\n");

    printf("AIB starting\n");
    {
      aib = vl_aib_new(Pic, nrows, ncols);
      vl_aib_process(aib);

      /*  parents always has size 2*nrows-1 */
      parents = vl_aib_get_parents(aib);
      for(r=0; r<2*nrows-1; r++)
        printf("%u => %u\n", r, parents[r]);

      vl_aib_delete(aib);
    }
    /* free(Pic); */

    printf("IB done\n");
    return 0;
}

File: toolbox/aib/aib.m
% AIB  Agglomerative Information Bottleneck
%   PARENTS = AIB(PCX) runs Agglomerative Information Bottleneck (AIB)
%   on the category-feature co-occurence matrix PCX and returns a
%   vector PARENTS representing all the fetures join operations.
%
%   The function iteratively merges the two vlaues of the feature X
%   that casuses the smallest decrease in mutual information between
%   the random variables X and C.
%
%   Merge operations are arranged in a binary tree. The nodes of the
%   tree correspond to the original feature values and any other value
%   obtained as a result of a merge operation. The nodes are numbered
%   in breadth-first order, starting from the leaves. The first node
%   has number one. The numbers associated to the leaves correspond to
%   the original feature values.  In total there are 2*M-1 nodes,
%   where M is the number of values (number of columns of PCX). The
%   internal nodes order reflects the order of the AIB merge
%   operations. It is therefore possible to recover from the tree the
%   state of the AIB algorithm at each step (see AIBCUT(),
%   AIBFINDCUT()).
%
%   The function returns a UINT32 array with one element per tree
%   node. Each element is the index of the parent node. The root
%   parent is equal to 1.
%
%   Feature values with null probability are ignored by the AIB
%   algorithm. By default these are exlcuded from the caluclation and
%   their parent is set to zero. Notice that this causes the root of
%   the tree to have index smaller of 2*M-1 (the returned vector is
%   still 2*M-1 elements lenght, but the last portion is
%   zero-padded).
%
%   Alternatively, the option ClusterNull can be used to cluster the
%   null nodes in its own cluster. In this case, the results is as if
%   pretenting that the null node have very small probability, uniform
%   across categories.
%
%   [PARENTS, COST] = AIB(...) returns the values cost of the COST
%   function being optimizied, one for each node in PARENTS. COST has
%   M column. The first column is the inital value of the cost
%   function. The others correspond to the cost after each of the
%   M-1 merges. If less than M-1 merges are performed, the rest of
%   the vector is filled with NaNs.
%
%   Options:
%
%   Verbose::
%     Increase verbosity level.
%
%   CLusterNull::
%     Do not signal null nodes; instead cluster them.
%
%   See also AIBCUT().

% AUTORIGHTS
% Copyright 2007 (c) Andrea Vedaldi and Brian Fulkerson
% 
% This file is part of VLFeat, available in the terms of the GNU
% General Public License version 2.
